commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/core/EuclideanNormAlgorithms.java

```diff
@@ -0,0 +1,751 @@
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.numbers.examples.jmh.core;
+
+import java.math.BigDecimal;
+import java.math.MathContext;
+import java.util.function.ToDoubleFunction;
+
+import org.apache.commons.numbers.core.Sum;
+
+/** Class containing various Euclidean norm computation methods for comparison.
+ */
+public final class EuclideanNormAlgorithms {
+
+    /** No instantiation. */
+    private EuclideanNormAlgorithms() {}
+
+    /** Exact computation method using {@link BigDecimal} and {@link MathContext#DECIMAL128}.
+     */
+    static final class Exact implements ToDoubleFunction<double[]> {
+
+        /** {@inheritDoc} */
+        @Override
+        public double applyAsDouble(final double[] v) {
+            // compute the sum of squares
+            final MathContext ctx = MathContext.DECIMAL128;
+
+            BigDecimal sum = BigDecimal.ZERO;
+            BigDecimal n;
+            for (int i = 0; i < v.length; ++i) {
+                n = Double.isFinite(v[i]) ? new BigDecimal(v[i]) : BigDecimal.ZERO;
+                sum = sum.add(n.multiply(n, ctx), ctx);
+            }
+
+            return sum.sqrt(ctx).doubleValue();
+        }
+    }
+
+    /** Direct computation method that simply computes the sums of squares and takes
+     * the square root with no special handling of values.
+     */
+    static final class Direct implements ToDoubleFunction<double[]> {
+
+        /** {@inheritDoc} */
+        @Override
+        public double applyAsDouble(final double[] v) {
+            double n = 0;
+            for (int i = 0; i < v.length; i++) {
+                n += v[i] * v[i];
+            }
+            return Math.sqrt(n);
+        }
+    }
+
+    /** Translation of the <a href="http://www.netlib.org/minpack">minpack</a>
+     * "enorm" subroutine. This method handles overflow and underflow.
+     */
+    static final class Enorm implements ToDoubleFunction<double[]> {
+
+        /** Constant. */
+        private static final double R_DWARF = 3.834e-20;
+        /** Constant. */
+        private static final double R_GIANT = 1.304e+19;
+
+        /** {@inheritDoc} */
+        @Override
+        public double applyAsDouble(final double[] v) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 118-131 into a separate method.